pkg/util/pki/csr.go

```diff
@@ -217,12 +217,18 @@ func GenerateCSR(crt *v1.Certificate) (*x509.CertificateRequest, error) {
 			asn1ExtendedUsages = append(asn1ExtendedUsages, oid)
 		}
 	}
-	extendedUsage := pkix.Extension{
-		Id: OIDExtensionExtendedKeyUsage,
-	}
-	extendedUsage.Value, err = asn1.Marshal(asn1ExtendedUsages)
-	if err != nil {
-		return nil, fmt.Errorf("failed to asn1 encode extended usages: %w", err)
+
+	extraExtensions := []pkix.Extension{usage}
+	if len(ekus) > 0 {
+		extendedUsage := pkix.Extension{
+			Id: OIDExtensionExtendedKeyUsage,
+		}
+		extendedUsage.Value, err = asn1.Marshal(asn1ExtendedUsages)
+		if err != nil {
+			return nil, fmt.Errorf("failed to asn1 encode extended usages: %w", err)
+		}
+
+		extraExtensions = append(extraExtensions, extendedUsage)
```

Comment:
Having read the code again I'd suggest that we create a helper function `extraUsagesForCertificate` which has all this logic and which can be unit tested separately. 
That would be consistent with the early sections of this function which call out to `IPAddressesForCertificate`, `SubjectForCertificate` etc.

Having that separate function would allow us to explicitly test the case where no extended usages are supplied in the Certificate.

Reply:
I also don't like how `BuildKeyUsages` does its own defaulting of the usages. 
I think we should start writing `Certificate` defaulting functions.
Ideally they would be called by top level reconciler function as soon as the Certificate has been fetched from the API server, so that all other code can safely asssume that default have been applied.

Ultimately, those same defaulting functions can be used in a defaulting webhook.

Reply:
Splitted up logic!

Re: defaulting 
would be a good thing to reconsider once we go beyond kubernetes 1.11